File: backend/app/models/transaction.py
"""Module for backend app models transaction."""

import enum
from datetime import date
from typing import TYPE_CHECKING

from sqlalchemy import Date, Enum, Float, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ..database import Base

if TYPE_CHECKING:
    from .classification import ClassificationSession, RecurrencePattern


class TransactionType(enum.Enum):
    """Represent transaction type."""

    INCOME = "Income"
    EXPENSE = "Expense"
    # Movements between owned accounts; excluded from income/expense totals.
    TRANSFER = "Transfer"


class ExpenseType(enum.Enum):
    """Represent expense type."""

    FIXED_ESSENTIAL = "Fixed Essential"  # Ramit's "Fixed Costs" (50-60%)
    GUILT_FREE_DISCRETIONARY = "Discretionary"  # Ramit's "Guilt-Free Spending" (20-35%)
    SAVINGS_INVESTMENT = (
        "Savings & Investment"  # Ramit's "Investments" & "Savings" (10-20%)
    )
    NEUTRAL = "Neutral"  # For non-spending operational categories


class ExpenseCategory(enum.Enum):
    """Represent expense category."""

    # --- 1. Fixed Essentials (Survival & Obligations) ---
    HOUSING = "Housing"  # Rent, Mortgage
    UTILITIES = "Utilities"  # Energy, Water, Internet
    GROCERIES = "Groceries"  # Supermarket food
    TRANSPORTATION = "Transportation"  # Public transport, Car insurance, Gas
    INSURANCE = "Insurance"  # Mandatory insurance (Zorgpremie, Family)
    HEALTH = "Health"  # Doctor, Pharmacy (Essential medical)

    # --- 2. Debt & Financial Obligations (Essential) ---
    LOAN_REPAYMENT = "Loan Repayment"  # Fixed loan payments (e.g., Ivan)
    CREDIT_PAYMENT = "Credit Payment"  # Repaying credit card/line (e.g., Belfius debt)
    DEBT = "Debt"  # General debt repayment
    FINANCIAL_FEES = "Financial Fees"  # Bank fees, late fees

    # --- 3. Savings & Investments (Future Wealth) ---
    INVESTMENTS = "Investments"  # ETFs, Crypto, Stocks
    SAVINGS = "Savings"  # Emergency Fund, specific saving goals

    # --- 4. Guilt-Free Discretionary (Lifestyle) ---
    EATING_OUT = "Eating Out"  # Restaurants, UberEats
    PERSONAL = "Personal"  # Gym (Stadium), Haircuts, Cosmetics
    SHOPPING = "Shopping"  # Clothes, Gadgets (Non-essential)
    GIFTS = "Gifts"
    DONATIONS = "Donations"
    EDUCATION = "Education"  # Courses, Books (Self-improvement)
    TRAVEL = "Travel"
    ENTERTAINMENT = "Entertainment"  # Movies, Netflix, Events

    # --- 5. Neutral/Operational ---
    OTHERS = "Others"

    @property
    def expense_type(self) -> ExpenseType:
        """Classify expense categories according to the spending plan."""
        # 50-60% of Income
        fixed_essential = [
            ExpenseCategory.HOUSING,
            ExpenseCategory.UTILITIES,
            ExpenseCategory.GROCERIES,
            ExpenseCategory.TRANSPORTATION,
            ExpenseCategory.INSURANCE,
            ExpenseCategory.HEALTH,
            ExpenseCategory.LOAN_REPAYMENT,
            ExpenseCategory.CREDIT_PAYMENT,
            ExpenseCategory.DEBT,
            ExpenseCategory.FINANCIAL_FEES,
        ]

        # 10-20% of Income
        savings_investments = [ExpenseCategory.INVESTMENTS, ExpenseCategory.SAVINGS]

        if self in fixed_essential:
            return ExpenseType.FIXED_ESSENTIAL
        if self in savings_investments:
            return ExpenseType.SAVINGS_INVESTMENT
        # All remaining are Guilt-Free Discretionary (20-35%)
        return ExpenseType.GUILT_FREE_DISCRETIONARY

    @property
    def is_essential(self) -> bool:
        """Handle is essential."""
        return self.expense_type == ExpenseType.FIXED_ESSENTIAL

    @property
    def is_discretionary(self) -> bool:
        """Handle is discretionary."""
        return self.expense_type == ExpenseType.GUILT_FREE_DISCRETIONARY

    @classmethod
    def get_essential_categories(cls) -> list["ExpenseCategory"]:
        """Return essential categories."""
        return [category for category in cls if category.is_essential]

    @classmethod
    def get_discretionary_categories(cls) -> list["ExpenseCategory"]:
        """Return discretionary categories."""
        return [category for category in cls if category.is_discretionary]


class IncomeCategory(enum.Enum):
    """Represent income category."""

    SALARY = "Salary"
    INVESTMENTS = "Investment Income"
    BUSINESS = "Business Income"
    RENTAL = "Rental Income"
    FREELANCE = "Freelance Income"
    PENSION = "Pension"
    BENEFITS = "Benefits"  # Unemployment, Child benefits
    GIFTS = "Gifts Received"
    REFUNDS = "Refunds"  # Tax returns, shop refunds
    LOAN_DISBURSEMENT = (
        "Loan Disbursement"  # Incoming money from Credit Line/Loans (NOT EARNINGS)
    )
    OTHER = "Other Income"


class TransferCategory(enum.Enum):
    """Represent transfer category."""

    INTERNAL_TRANSFER = "Internal Transfer"
    CREDIT_CARD_SETTLEMENT = "Credit Card Settlement"
    LOAN_TO_PERSON = "Loan to Person"
    LOAN_REPAYMENT_RECEIVED = "Loan Repayment Received"
    LOAN_FROM_PERSON = "Loan from Person"
    DEBT_REPAYMENT_SENT = "Debt Repayment Sent"


class Transaction(Base):
    """Represent transaction."""

    __tablename__ = "transactions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    import_session_id: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("import_sessions.id"), index=True, nullable=True
    )
    import_source_locator: Mapped[str | None] = mapped_column(
        String(255), nullable=True
    )
    import_source_description: Mapped[str | None] = mapped_column(
        String(500), nullable=True
    )
    canonical_description_en: Mapped[str | None] = mapped_column(
        String(500), nullable=True
    )
    account_number: Mapped[str] = mapped_column(String(50), index=True, nullable=True)
    transaction_date: Mapped[date] = mapped_column(Date, index=True, nullable=True)
    amount: Mapped[float] = mapped_column(Float, nullable=True)
    currency: Mapped[str] = mapped_column(String(3), nullable=True)
    description: Mapped[str] = mapped_column(String(500), nullable=True)
    counterparty_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    counterparty_account: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Updated Enum columns
    transaction_type: Mapped[TransactionType] = mapped_column(
        Enum(TransactionType), nullable=True
    )
    expense_category: Mapped[ExpenseCategory | None] = mapped_column(
        Enum(ExpenseCategory), nullable=True
    )
    income_category: Mapped[IncomeCategory | None] = mapped_column(
        Enum(IncomeCategory), nullable=True
    )
    transfer_category: Mapped[TransferCategory | None] = mapped_column(
        Enum(TransferCategory), nullable=True
    )
    classification_source: Mapped[str | None] = mapped_column(String(50), nullable=True)
    recurrence_pattern_id: Mapped[int | None] = mapped_column(
        Integer,
        ForeignKey(
            "recurrence_patterns.id",
            name="fk_transactions_recurrence_pattern_id",
            use_alter=True,
        ),
        nullable=True,
        index=True,
    )

    source_bank: Mapped[str] = mapped_column(String(10), nullable=True)

    classification_sessions: Mapped[list["ClassificationSession"]] = relationship(
        "ClassificationSession",
        back_populates="transaction",
        cascade="all, delete-orphan",
    )
    recurrence_pattern: Mapped["RecurrencePattern | None"] = relationship(
        "RecurrencePattern", foreign_keys=[recurrence_pattern_id], post_update=True
    )
    seeded_recurrence_patterns: Mapped[list["RecurrencePattern"]] = relationship(
        "RecurrencePattern",
        foreign_keys="RecurrencePattern.seed_transaction_id",
        back_populates="seed_transaction",
        cascade="all, delete-orphan",
    )

File: frontend/src/types/transaction.ts
export enum TransactionType {
  INCOME = "Income",
  EXPENSE = "Expense",
  TRANSFER = "Transfer"
}

export enum TransferCategory {
  INTERNAL_TRANSFER = "Internal Transfer",
  CREDIT_CARD_SETTLEMENT = "Credit Card Settlement",
  LOAN_TO_PERSON = "Loan to Person",
  LOAN_REPAYMENT_RECEIVED = "Loan Repayment Received",
  LOAN_FROM_PERSON = "Loan from Person",
  DEBT_REPAYMENT_SENT = "Debt Repayment Sent"
}

export enum ExpenseType {
  FIXED_ESSENTIAL = "Fixed Essential",
  GUILT_FREE_DISCRETIONARY = "Discretionary",
  SAVINGS_INVESTMENT = "Savings & Investment",
  NEUTRAL = "Neutral"
}

export enum ExpenseCategory {
  // Fixed Essentials (Survival & Obligations)
  HOUSING = "Housing",
  UTILITIES = "Utilities",
  GROCERIES = "Groceries",
  TRANSPORTATION = "Transportation",
  INSURANCE = "Insurance",
  HEALTH = "Health",
  
  // Debt & Financial Obligations (Essential)
  LOAN_REPAYMENT = "Loan Repayment",
  CREDIT_PAYMENT = "Credit Payment",
  DEBT = "Debt",
  FINANCIAL_FEES = "Financial Fees",
  
  // Savings & Investments (Future Wealth)
  INVESTMENTS = "Investments",
  SAVINGS = "Savings",
  
  // Guilt-Free Discretionary (Lifestyle)
  EATING_OUT = "Eating Out",
  PERSONAL = "Personal",
  SHOPPING = "Shopping",
  GIFTS = "Gifts",
  DONATIONS = "Donations",
  EDUCATION = "Education",
  TRAVEL = "Travel",
  ENTERTAINMENT = "Entertainment",
  
  OTHERS = "Others"
}

export enum IncomeCategory {
  SALARY = "Salary",
  INVESTMENTS = "Investment Income",
  BUSINESS = "Business Income",
  RENTAL = "Rental Income",
  FREELANCE = "Freelance Income",
  PENSION = "Pension",
  BENEFITS = "Benefits",
  GIFTS = "Gifts Received",
  REFUNDS = "Refunds",
  LOAN_DISBURSEMENT = "Loan Disbursement",
  OTHER = "Other Income"
}

export interface DisplayMoneyFields {
  display_amount?: number | null;
  display_currency?: string | null;
  display_fx_rate?: number | null;
  display_rate_date?: string | null;
  display_is_available?: boolean | null;
  display_unavailable_reason?: string | null;
}

export interface Transaction extends DisplayMoneyFields {
  id: number;
  account_number: string;
  transaction_date: string;
  amount: number;
  currency: string;
  description: string;
  counterparty_name?: string;
  counterparty_account?: string;
  transaction_type: TransactionType;
  expense_category?: ExpenseCategory;
  income_category?: IncomeCategory;
  transfer_category?: TransferCategory;
  classification_source?:
    | 'manual'
    | 'assistant'
    | 'assistant_batch'
    | 'upload_suggester'
    | 'recurrence_pattern'
    | null;
  recurrence_pattern_id?: number | null;
  source_bank: string;
}

export interface CategoryStatistics {
  category: string;
  transaction_type: TransactionType;
  period?: string;
  date?: string;
  
  // For backward compatibility
  total_amount: number;
  transaction_count: number;
  
  // Period-specific metrics
  period_amount?: number;
  period_transaction_count?: number;
  period_percentage?: number;
  
  // Cumulative metrics
  total_amount_cumulative?: number;
  total_transaction_count?: number;
  
  // Averages
  average_transaction_amount?: number;
  
  // Yearly metrics
  yearly_amount?: number;
  yearly_transaction_count?: number;
}

export interface CategorySuggestion {
  category: string;
  confidence: number;
}

export interface SuggestCategoryResponse {
  suggestions: CategorySuggestion[];
}

export interface SortParams {
  field: 'date' | 'description' | 'amount' | 'type';
  direction: 'asc' | 'desc';
}

export type ClassificationStatusFilter = 'all' | 'classified' | 'unclassified';

export enum ActionType {
  DELETE_TRANSACTION = 'DELETE_TRANSACTION',
  UPDATE_CATEGORY = 'UPDATE_CATEGORY'
}

export enum TimePeriod {
  THREE_MONTHS = "3M",
  SIX_MONTHS = "6M",
  YEAR_TO_DATE = "YTD",
  ONE_YEAR = "1Y",
  TWO_YEARS = "2Y",
  ALL_TIME = "ALL_TIME"
}

export interface DeleteTransactionAction {
  type: ActionType.DELETE_TRANSACTION;
  transaction: Transaction;
}

export interface UpdateCategoryAction {
  type: ActionType.UPDATE_CATEGORY;
  transactionId: number;
  oldCategory: ExpenseCategory | IncomeCategory | TransferCategory | undefined;
  newCategory: ExpenseCategory | IncomeCategory | TransferCategory;
  oldTransactionType: TransactionType;
  newTransactionType: TransactionType;
}

export type UndoableAction = DeleteTransactionAction | UpdateCategoryAction;

export interface WeekdayStats {
  count: number;
  total: number;
  average: number;
  median: number;
  min: number;
  max: number;
}

export interface WeekdayTypeStats {
  expense: WeekdayStats;
  income: WeekdayStats;
}

export interface WeekdayDistribution {
  weekdays: {
    [key: string]: WeekdayTypeStats;
  };
  transaction_count: number;
}

// Financial Health Types
export interface FinancialHealthScore {
  id: number;
  date: string;
  overall_score: number;
  savings_rate_score: number;
  expense_ratio_score: number;
  budget_adherence_score: number;
  debt_to_income_score: number;
  emergency_fund_score: number;
  spending_stability_score: number;
  investment_rate_score: number;
  
  // Raw metrics
  savings_rate: number;
  expense_ratio: number;
  budget_adherence: number;
  debt_to_income: number;
  emergency_fund_months: number;
  spending_stability: number;
  investment_rate: number;
  
  // Recommendations
  recommendations?: RecommendationData[];
}

export interface FinancialHealthHistory {
  dates: string[];
  overall_scores: number[];
  savings_rate_scores: number[];
  expense_ratio_scores: number[];
  budget_adherence_scores: number[];
  debt_to_income_scores: number[];
  emergency_fund_scores: number[];
  spending_stability_scores: number[];
  investment_rate_scores: number[];
}

export interface RecommendationData {
  title: string;
  description: string;
  category: string;
  impact_area: string;
  priority: number;
  estimated_score_improvement: number;
}

export interface Recommendation {
  id: number;
  title: string;
  description: string;
  category: string;
  impact_area: string;
  priority: number;
  estimated_score_improvement: number;
  is_completed: boolean;
  date_completed: string | null;
  date_created: string;
}
